### ISSUE TITLE
[Bug] Maximum Call Stack Exceeded when  referencing recursive object

### ISSUE THREAD

**Issue by BrianUribe6** [valtiojs/valtio-yjs#46]
Hello,
First of all awesome library it has saved me countless hours of effort when integrating Yjs and React.

I've recently ran into an issue when attempting to use a **tree-like** state.  My state object looks something like this:

```ts 
type  Item = {
 parent: Item | Item[]  // <--- This causes infinite recursion when binding
 label: string;
 children: Item[]
}

const doc = new Y.Doc();
const state = proxy<Item[]>([]);

bind(state, doc.getArray("sidebar");

function addItem(parent: Item | Item[], item: Item) {
  if (Array.isArray(parent)) {
    parent.push(item);
  } else {
    parent.children.push(item);
   }
   item.parent = parent;
}
```

I need to keep track of the `parent` of each item, and also allow adding child elements to any Item. 

`bind` seems to infinitely recurse when trying to access  the `parent` .

I've a created a minimal reproduction:

https://codesandbox.io/p/sandbox/valtio-yjs-recursive-type-bug-c3hjhh?file=%2Fsrc%2FApp.js%3A14%2C1

Thank you!

**Comment by raineorshine** [comment]
valtio-yjs has to convert the proxy into a valid YJS structure to work. Is there a valid YJS structure with circular references? I get an error when I try it:

```js
import * as Y from 'yjs'

const doc = new Y.Doc()
const root = doc.getMap()
const parent = new Y.Map()
const child = new Y.Map()

root.set('parent', parent)
parent.set('child', child)

// ERROR: Cannot read properties of null (reading 'forEach')
child.set('parent', parent)

console.log(doc.toJSON())
```

Demo: [View in CodeSandbox](https://codesandbox.io/p/devbox/yjs-ult-and-penult-forked-64j2h9?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clr18dzrz00063b6b3noddegn%2522%252C%2522sizes%2522%253A%255B67.81111353898478%252C32.188886461015215%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clr18dzrz00023b6b44af2bng%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clr18dzrz00043b6b386hgn4h%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clr18dzrz00053b6bpdcm6hi4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B100%252C0%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clr18dzrz00023b6b44af2bng%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clr18dzrz00013b6b09bplb1v%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage-lock.json%2522%257D%255D%252C%2522id%2522%253A%2522clr18dzrz00023b6b44af2bng%2522%252C%2522activeTabId%2522%253A%2522clr18dzrz00013b6b09bplb1v%2522%257D%252C%2522clr18dzrz00053b6bpdcm6hi4%2522%253A%257B%2522id%2522%253A%2522clr18dzrz00053b6bpdcm6hi4%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clr18dzrz00043b6b386hgn4h%2522%253A%257B%2522id%2522%253A%2522clr18dzrz00043b6b386hgn4h%2522%252C%2522activeTabId%2522%253A%2522clr18eafp006p3b6bx5r20zbe%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clr18dzrz00033b6bzzsykpxc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clr18e0900003egcxhbl71ers%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clr18eafp006p3b6bx5r20zbe%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Afalse%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by BrianUribe6** [comment]
> valtio-yjs has to convert the proxy into a valid YJS structure to work. Is there a valid YJS structure with circular references? I get an error when I try it:
> 
> ```js
> import * as Y from 'yjs'
> 
> const doc = new Y.Doc()
> const root = doc.getMap()
> const parent = new Y.Map()
> const child = new Y.Map()
> 
> root.set('parent', parent)
> parent.set('child', child)
> 
> // ERROR: Cannot read properties of null (reading 'forEach')
> child.set('parent', parent)
> 
> console.log(doc.toJSON())
> ```
> 
> https://codesandbox.io/p/devbox/yjs-ult-and-penult-forked-64j2h9

Interesting,  I modified your example like this. It still produces the same error:

```ts 
import * as Y from 'yjs'

const doc = new Y.Doc();
const root = doc.getMap();
const child = new Y.Map();

root.set("child", child);
child.set("parent", root); // <- Error on circular reference

console.log(child)
```

Does this mean that Yjs does not support any data structure with circular references? or is this a bug?

**Comment by PhilGarb** [comment]
I think this has to do with yjs itself. It is not possible to insert a shared type twice in a document like the author explained here: https://github.com/yjs/yjs/issues/458#issuecomment-1224800219. If only the data is relevant you could: `root.clone()`. I am not sure that circular references are even supported by yjs.

**Comment by raineorshine** [comment]
The solution is to assign a unique id to each Item, and reference the parent and children by id, i.e. foreign key. Then you can store all items in a single Map. 

For an even more normalized structure, you can store nodes and edges in separate Maps.

**Comment by BrianUribe6** [comment]
Thanks everyone for your assistance. I encountered this issue after doing some refactoring on my code. Previously, I used to pass the `child` and ` parent` references as separate values around the application. I thought it would be a good idea to combine them so that I could do `child.parent` instead, but now I understand why it is not possible.

> The solution is to assign a unique id to each Item, and reference the parent and children by id, i.e. foreign key. Then you can store all items in a single Map.

I will be taking @raineorshine  suggestions, since having to separately keep track of every node and its parent is very repetitive. 

I'll mark this as closed.  Once again thank you all!
